### PR TITLE
feat(DB): Add methods to get_dbc_attribute

### PR DESCRIPTION
**Pull request by d-bohls**
Add get_dbc_attribute method for cluster,
frame, ECU, and signal objects.

Add tests and a DBC containing attributes.

- [X] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nixnet-python/blob/master/CONTRIBUTING.rst).
- [X] New tests have been created for any new features or regression tests for bugfixes.
- [X] `tox` successfully runs, including unit tests and style checks (see [CONTRIBUTING.md](https://github.com/ni/nixnet-python/blob/master/CONTRIBUTING.rst)).

### CODE REVIEW

**Comment by d-bohls** [comment]
I now see that integration tests don't affect code coverage. I'm going to add a unit test.

**Comment by epage** [comment]
We track integration coverage separately from unit testing coverage. The combination of the two is the important thing.
________________________________

File: nixnet/database/_cluster.py [hunk 1]
```diff
@@ -48,6 +49,16 @@ def __hash__(self):
     def __repr__(self):
         return 'Cluster(handle={0})'.format(self._handle)
 
+    def get_dbc_attribute(
```
[the remark this answers is not in the document]
Comment (d-bohls) [reply nixnet/database/_cluster.py]: I like these suggestions. I'll have an updated pull request up shortly.

Regarding get_signal_value_table, we talked in person about how I currently can't find a DBC database containing a signal value table. I'll need one to actually test with if I'm going to be manipulating the string result from nxdbGetDBCAttribute. It seems promising I may be able to get one soon. I'll likely create an issue for that method and we can add it once I get the DBC.

File: nixnet/database/_dbc_attributes.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+    def keys(self):
+        mode = constants.GetDbcAttributeMode.ATTRIBUTE_LIST
+        attribute_size = _funcs.nxdb_get_dbc_attribute_size(self._handle, mode, '')
+        attribute_info = _funcs.nxdb_get_dbc_attribute(self._handle, mode, '', attribute_size)
+        name_string = attribute_info[0]
+        name_list = []  # type: typing.List[typing.Text]
+        if len(name_string):
+            name_list = name_string.split(',')
+        for name in name_list:
```
Comment (jashnani) [inline nixnet/database/_dbc_attributes.py]: No need to create a list. Can be simplified to:
`for name in name_string.split(",")`
  `if name.strip():`

File: tests/test_dbc_attributes.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        # check values
+        assert sig.dbc_attributes['mock_name'] == ('', False)
+
+
+@pytest.mark.integration
+def test_dbc_attributes_for_cluster():
+    database_filepath = os.path.join(os.path.dirname(__file__), 'databases\\attributes.dbc')
+    with database.Database(database_filepath) as db:
+        cluster = db.clusters['Cluster']
+        frame1 = cluster.frames['Msg1']
```
[the remark this answers is not in the document]
Comment (d-bohls) [reply tests/test_dbc_attributes.py]: I'm using the frame just to have another db object with dbc attribute's to use in some of the asserts below that exercise the dunders. The DBC database has only one cluster, and I can't easily edit the DBC file to add another.

File: tests/test_dbc_attributes.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+        # check dunders
+        assert len(cluster.dbc_attributes) == 2
+        assert len({cluster.dbc_attributes, cluster.dbc_attributes}) == 1
+        assert len({cluster.dbc_attributes, frame1.dbc_attributes}) == 2
+        assert cluster.dbc_attributes == cluster.dbc_attributes
+        assert cluster.dbc_attributes != frame1.dbc_attributes
```
[the remark this answers is not in the document]
Comment (d-bohls) [reply tests/test_dbc_attributes.py]: see response above

File: nixnet/database/_cluster.py [hunk 5]
```diff
@@ -20,6 +21,7 @@ class Cluster(object):
 
     def __init__(self, handle):
         self._handle = handle
+        self._dbc_attributes = _dbc_attributes.DbcAttributeCollection(handle)
```
[the remark this answers is not in the document]
Comment (d-bohls) [reply nixnet/database/_cluster.py]: Done. Fixed other spots as well.

File: nixnet/database/_dbc_attributes.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+
+class DbcAttributeCollection(collections.Mapping):
+
+    def __init__(self, handle):
+        # type: (int) -> None
+        self._handle = handle
+
+    def __repr__(self):
+        return 'DbcAttributeCollection(handle={0})'.format(self._handle)
```
[the remark this answers is not in the document]
Comment (d-bohls) [reply nixnet/database/_dbc_attributes.py]: Done.

File: nixnet/database/_dbc_attributes.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+    def __len__(self):
+        return len(list(self.keys()))
+
+    def __iter__(self):
+        return self.keys()
+
+    def __getitem__(self, index):
+        # type: (typing.Text) -> typing.Tuple[typing.Text, bool]
+        """Get attribute properties (value and is default value) for a DBC attribute.
```
[the remark this answers is not in the document]
Comment (d-bohls) [reply nixnet/database/_dbc_attributes.py]: Done

File: nixnet/database/_dbc_attributes.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+            Args:
+                index(str): attribute name.
+            Returns:
+                tuple(str, bool): attribute value and is default value.
```
[the remark this answers is not in the document]
Comment (d-bohls) [reply nixnet/database/_dbc_attributes.py]: Done

File: nixnet/database/_dbc_attributes.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        """
+        if isinstance(index, six.string_types):
+            return self._value(index)
+        else:
+            raise TypeError(index)
+
+    def keys(self):
+        """Returns all the keys (attribute names) in the DbcAttributeCollection object.
```
[the remark this answers is not in the document]
Comment (d-bohls) [reply nixnet/database/_dbc_attributes.py]: Done

File: nixnet/database/_dbc_attributes.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+        attribute_info = _funcs.nxdb_get_dbc_attribute(self._handle, mode, '', attribute_size)
+        name_string = attribute_info[0]
+        name_list = []
+        if len(name_string):
+            name_list = name_string.split(',')
+        for name in name_list:
+            yield name
+
+    def values(self):
+        """Returns all the values (attribute properties) in the DbcAttributeCollection object.
```
[the remark this answers is not in the document]
Comment (d-bohls) [reply nixnet/database/_dbc_attributes.py]: Done

File: nixnet/database/_dbc_attributes.py [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+            Yields:
+                An iterator to all the values (attribute properties) in the DbcAttributeCollection object.
+        """
+        names = list(self.keys())
+        for name in names:
+            yield self._value(name)
+
+    def items(self):
+        """Returns all the attributes names and properties in the DbcAttributeCollection object.
```
[the remark this answers is not in the document]
Comment (d-bohls) [reply nixnet/database/_dbc_attributes.py]: Done.

Please double check keys/values/items documentation. I figured you wanted the "Yields" part to more closely match AliasCollection as well.

File: tests/test_dbc_attributes.py [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+MockXnetLibrary = mock.create_autospec(_cfuncs.XnetLibrary, spec_set=True, instance=True)
+MockXnetLibrary.nxdb_open_database.return_value = _ctypedefs.u32(0)
+MockXnetLibrary.nxdb_close_database.return_value = _ctypedefs.u32(0)
+MockXnetLibrary.nxdb_find_object.return_value = _ctypedefs.u32(0)
+MockXnetLibrary.nxdb_get_dbc_attribute_size.return_value = _ctypedefs.u32(0)
+MockXnetLibrary.nxdb_get_dbc_attribute.return_value = _ctypedefs.u32(0)
+
+
+@mock.patch('nixnet._cfuncs.lib', MockXnetLibrary)
+def test_dbc_attributes_mock():
```
[the remark this answers is not in the document]
Comment (d-bohls) [reply tests/test_dbc_attributes.py]: Done

File: tests/test_dbc_attributes.py [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+    attributes130 = _dbc_attributes.DbcAttributeCollection(130)
+    assert str(attributes120) == 'DbcAttributeCollection(handle=120)'
+    assert attributes120 == attributes120
+    assert attributes120 != attributes130
+    assert len({attributes120, attributes120}) == 1
+    assert len({attributes120, attributes130}) == 2
+
+
+@mock.patch('nixnet._cfuncs.lib', MockXnetLibrary)
+def test_dbc_attributes_for_cluster_mock():
```
[the remark this answers is not in the document]
Comment (d-bohls) [reply tests/test_dbc_attributes.py]: Done.

For now, we agreed to leave "_mock" at the end to distinguish unit test from the integration counterparts. Let me know if there is a better suggestion.

File: nixnet/database/_dbc_attributes.py [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+    def _value(self, name):
+        # type: (typing.Text) -> typing.Tuple[typing.Text, bool]
+        mode = constants.GetDbcAttributeMode.ATTRIBUTE
+        attribute_size = _funcs.nxdb_get_dbc_attribute_size(self._handle, mode, name)
+        attribute_info = _funcs.nxdb_get_dbc_attribute(self._handle, mode, name, attribute_size)
+        try:
+            # If this is an enum attribute, replace the enum index with the enum string.
+            # First, see if the attribute value can be cast to a positive int.
+            # If the attribute value has no chance of being an index, we can avoid
+            # calling the C API to get the enum list.
```
[the remark this answers is not in the document]
Comment (d-bohls) [reply nixnet/database/_dbc_attributes.py]: Done

File: nixnet/database/_dbc_attributes.py [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+            value = attribute_info[0]
+            index = int(value)
+            if index >= 0:
+                # The attribute value might represent an enum index.
+                # Go ahead and call into the C API to get the list of enums.
+                enums = self._enums(name)
+                if index < len(enums):
+                    # Repackage the tuple with the enum string and is_default
+                    attribute_info = enums[index], attribute_info[1]
+        finally:
```
[the remark this answers is not in the document]
Comment (d-bohls) [reply nixnet/database/_dbc_attributes.py]: Done.

I don't expect any errors other than ValueError. I limited the scope of the try to just the casting.

File: nixnet/database/_dbc_attributes.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    def __init__(self, handle):
+        # type: (int) -> None
+        self._handle = handle
+
+    def __repr__(self):
+        return 'DbcAttributeCollection(handle={0})'.format(self._handle)
+
+    def __eq__(self, other):
+        if isinstance(other, self.__class__):
+            return self._handle == other._handle
```
[the remark this answers is not in the document]
Comment (d-bohls) [reply nixnet/database/_dbc_attributes.py]: Done

File: nixnet/database/_ecu.py [hunk 17]
```diff
@@ -46,6 +50,10 @@ def comment(self, value):
     def config_status(self):
         return _props.get_ecu_config_status(self._handle)
 
+    @property
+    def dbc_attributes(self):
```
[the remark this answers is not in the document]
Comment (d-bohls) [reply nixnet/database/_ecu.py]: I checked them; they look good.

File: nixnet/database/_dbc_attributes.py [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+    def __init__(self, handle):
+        # type: (int) -> None
+        self._handle = handle
+
+        # Here, we are caching the attribute names and enums to work around a driver bug.
+        # The bug results in an empty attribute value after intermixing calls to get attribute values and enums.
+        # We can avoid the bug if we get all attribute enums first, before getting any attribute values.
+        self._cache = {}  # type: typing.Dict[typing.Text, typing.List[typing.Text]]
+        for name in self._get_names():
+            self._cache[name] = self._get_enums(name)
```
Comment (jashnani) [inline nixnet/database/_dbc_attributes.py]: You could make this more "pythonic" by doing the following.

self._cache =dict(
   (name, self._get_enums(name))
  for name in self._get_names)
)

File: tests/test_dbc_attributes.py [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+        ecu = cluster.ecus['mock_ecu']
+        frame = cluster.frames['mock_frame']
+        sig = frame.mux_static_signals['mock_signal']
+        results.append(cluster.get_dbc_attribute('mock_name'))
+        results.append(ecu.get_dbc_attribute('mock_name', constants.GetDbcAttributeMode.ATTRIBUTE))
+        results.append(frame.get_dbc_attribute('mock_name', constants.GetDbcAttributeMode.ATTRIBUTE_LIST))
+        results.append(sig.get_dbc_attribute('mock_name', constants.GetDbcAttributeMode.VALUE_TABLE_LIST))
+        results.append(sig.get_dbc_attribute('mock_name', constants.GetDbcAttributeMode.ENUMERATION_LIST))
+
+    assert results == expected_results
```
[the remark this answers is not in the document]
Comment (jashnani) [reply tests/test_dbc_attributes.py]: Fixed.

File: tests/test_dbc_attributes.py [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+    expected_results = [
+        ('', False),
+        ('', False),
+        ('', False),
+        ('', False),
+        ('', False)]
+
+    with database.Database('mock_database') as db:
+        cluster = db.clusters['mock_cluster']
+        ecu = cluster.ecus['mock_ecu']
```
[the remark this answers is not in the document]
Comment (jashnani) [reply tests/test_dbc_attributes.py]: Fixed. Thanks for doing this!

File: nixnet/database/_cluster.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jashnani) [reply nixnet/database/_cluster.py]: Fixed. #226 created for addressing signal value table portion of DBC attributes.

File: nixnet/database/_cluster.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jashnani) [reply nixnet/database/_cluster.py]: Fixed.

File: nixnet/database/_dbc_attributes.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jashnani) [reply nixnet/database/_dbc_attributes.py]: Fixed.

File: nixnet/database/_dbc_attributes.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jashnani) [reply nixnet/database/_dbc_attributes.py]: Fixed.

File: nixnet/database/_dbc_attributes.py [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+        attribute_size = _funcs.nxdb_get_dbc_attribute_size(self._handle, mode, name)
+        attribute_info = _funcs.nxdb_get_dbc_attribute(self._handle, mode, name, attribute_size)
+        try:
+            # If this is an enum attribute, replace the enum index with the enum string.
+            # First, see if the attribute value can be cast to a positive int.
+            # If the attribute value has no chance of being an index, we can avoid
+            # calling the C API to get the enum list.
+            value = attribute_info[0]
+            index = int(value)
+            if index >= 0:
```
[the remark this answers is not in the document]
Comment (jashnani) [reply nixnet/database/_dbc_attributes.py]: No longer an issue as you changed the design to use a cache.

File: tests/test_dbc_attributes.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (jashnani) [reply tests/test_dbc_attributes.py]: We decided to replace these with one unit test for `_dbc_attributes` class. No longer an issue.

File: nixnet/database/_dbc_attributes.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jashnani) [reply nixnet/database/_dbc_attributes.py]: Fixed.

File: nixnet/database/_dbc_attributes.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jashnani) [reply nixnet/database/_dbc_attributes.py]: Fixed.

File: nixnet/database/_dbc_attributes.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jashnani) [reply nixnet/database/_dbc_attributes.py]: Fixed.

File: nixnet/database/_dbc_attributes.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jashnani) [reply nixnet/database/_dbc_attributes.py]: Fixed.

File: nixnet/database/_dbc_attributes.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (jashnani) [reply nixnet/database/_dbc_attributes.py]: Fixed.

File: tests/test_dbc_attributes.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jashnani) [reply tests/test_dbc_attributes.py]: Replaced by one unit test. Fixed.

File: .gitignore [hunk 22]
```diff
@@ -1,6 +1,7 @@
 .cache/
 .mypy_cache/
 .tox/
+docs/_build/
```
[the remark this answers is not in the document]
Comment (jashnani) [reply .gitignore]: Fixed.

File: nixnet/database/_dbc_attributes.py (hunk 18 above)
Comment (jashnani) [reply nixnet/database/_dbc_attributes.py]: Fixed.

**Pull request merged**